### PR TITLE
Complete local exception refactoring in C++ and derived language mappings

### LINKED ISSUE #2402: Make LocalException public instead of open in Swift
`public class` seems more correct now. I suspect we needed `open` in 3.7 for local slice.

_Originally posted by @externl in https://github.com/zeroc-ice/ice/pull/2378#discussion_r1664206309_

### PR DESCRIPTION
**Pull request by bernardnormier**
This PR completes the local exception refactoring in C++ and derived language mappings.

Fixes #2402.

### CODE REVIEW

File: CHANGELOG-3.8.md [hunk 1]
```diff
@@ -59,36 +59,39 @@ These are the changes since the Ice 3.7.10 release in [CHANGELOG-3.7.md](./CHANG
...
+  | UnsupportedEncodingException        | MarshalException           |          |
+  | UnsupportedProtocolException        | MarshalException, FeatureNotSupportedException | |
+  | VersionMismatchException            | InitializationException    |          |
+  | VersionParseException               | ParseException             |          |
 
   base = was existing base class
 
   New local exceptions:\
-  ConnectionAbortedException, ConnectionClosedException, ConnectionIdleException, ParseException
+  ConnectionAbortedException, ConnectionClosedException, ParseException
```
Comment (bernardnormier) [inline CHANGELOG-3.8.md:94]: Replaced ConnectionIdleException by ConnectionAbortedException(closedByApplication = false).

File: cpp/include/Ice/LocalException.h [hunk 2]
```diff
@@ -16,7 +16,13 @@ namespace Ice
     class ICE_API LocalException : public Exception
     {
     public:
-        using Exception::Exception;
+        /**
```
Comment (bernardnormier) [inline cpp/include/Ice/LocalException.h:19]: All derived local exceptions must provide a message, meaning that for local exceptions, ice_id() is never used as the fallback message.

File: cpp/include/Ice/LocalExceptions.h [hunk 3]
```diff
@@ -323,897 +280,636 @@ namespace Ice
...
+     * This exception is raised if a system error occurred in the server or client process. There are many possible
+     * causes for such a system exception. For details on the cause, {@link SyscallException#error} should be inspected.
+     * \headerfile Ice/Ice.h
      */
-    class ICE_API NotRegisteredException : public LocalException
+    class ICE_API SyscallException : public LocalException
     {
     public:
-        using LocalException::LocalException;
+        // The error code is a DWORD (unsigned long) on Windows and an int on other platforms.
```
Comment (bernardnormier) [inline cpp/include/Ice/LocalExceptions.h:368]: That's another benefit of dropping local Slice: the error code for Windows can have the correct type.

File: cpp/src/Ice/Service.cpp [hunk 4]
```diff
@@ -176,7 +176,7 @@ namespace
             // Don't need to use a wide string converter as the wide string is passed
             // to Windows API.
             //
-            LONG err = RegCreateKeyExW(
+            LSTATUS err = RegCreateKeyExW(
```
Comment (bernardnormier) [inline cpp/src/Ice/Service.cpp:179]: The Windows API is weird: some error codes that can be "printed" with FormatMessage are not DWORDs.

File: python/modules/IcePy/Util.cpp [hunk 5]
```diff
@@ -727,48 +730,58 @@ IcePy::convertException(std::exception_ptr exPtr)
         rethrow_exception(exPtr);
     }
     // First handle exceptions with extra fields we want to provide to Python users.
-    catch (const Ice::RequestFailedException& ex)
+    catch (const Ice::AlreadyRegisteredException& ex)
     {
-        std::array<IcePy::PyObjectHandle, 4> args = {
```
Comment (bernardnormier) [inline python/modules/IcePy/Util.cpp:732]: It's much more elegant and less error prone for N to be inferred.

File: swift/test/Ice/exceptions/AllTests.swift [hunk 6]
```diff
@@ -14,8 +14,6 @@ class ServantLocatorI: Ice.ServantLocator {
     func deactivate(_: String) {}
 }
 
-class CustomLocalException: LocalException {}
```
Comment (bernardnormier) [inline swift/test/Ice/exceptions/AllTests.swift:17]: No longer worked now that LocalException is public instead of open.

File: CHANGELOG-3.8.md [hunk 7]
```diff
@@ -59,36 +59,39 @@ These are the changes since the Ice 3.7.10 release in [CHANGELOG-3.7.md](./CHANG
...
-  | UnknownMessageException             | ProtocolException (base)   |
-  | UnknownReplyStatusException         | MarshalException           |
-  | UnmarshalOutOfBoundsException       | MarshalException (base)    |
-  | UnsupportedEncodingException        | MarshalException           |
-  | UnsupportedProtocolException        | MarshalException, FeatureNotSupportedException |
-  | VersionParseException               | ParseException             |
+  | Local exception in Ice 3.7          | Replacement                | Notes    |
+  |-------------------------------------|----------------------------| ---------|
+  | BadMagicException                   | ProtocolException (base)   |          |
+  | CFNetworkException                  | SocketsException (base)    |          |
```
Comment (externl) [inline CHANGELOG-3.8.md:65]: Typo?
```suggestion
  | CFNetworkException                  | SocketException (base)    |          |
```

File: CHANGELOG-3.8.md [hunk 8]
```diff
@@ -59,36 +59,39 @@ These are the changes since the Ice 3.7.10 release in [CHANGELOG-3.7.md](./CHANG
...
+  | CloneNotImplementedException        | logic_error                | C++ only |
```
Comment (externl) [inline CHANGELOG-3.8.md:66]:
```suggestion
  | CloneNotImplementedException        | std::logic_error                | C++ only |
```

File: CHANGELOG-3.8.md [hunk 9]
```diff
@@ -59,36 +59,39 @@ These are the changes since the Ice 3.7.10 release in [CHANGELOG-3.7.md](./CHANG
...
+  | CompressionException                | ProtocolException (base)   |          |
+  | ConnectionManuallyClosedException   | ConnectionAbortedException, ConnectionClosedException | |
+  | ConnectionNotValidatedException     | ProtocolException (base)   |          |
+  | EncapsulationException              | MarshalException (base)    |          |
+  | EndpointParseException              | ParseException             |          |
+  | EndpointSelectionTypeParseException | ParseException             |          |
+  | IllegalIdentityException            | ArgumentException (C#), invalid_argument (C++)     | |
```
Comment (externl) [inline CHANGELOG-3.8.md:73]:
```suggestion
  | IllegalIdentityException            | ArgumentException (C#), std::invalid_argument (C++)     | |
```

File: CHANGELOG-3.8.md [hunk 10]
```diff
@@ -59,36 +59,39 @@ These are the changes since the Ice 3.7.10 release in [CHANGELOG-3.7.md](./CHANG
...
+  | IllegalMessageSizeException         | MarshalException           |          |
+  | IllegalServantException             | ArgumentNullException (C#), invalid_argument (C++) | |
```
Comment (externl) [inline CHANGELOG-3.8.md:75]:
```suggestion
  | IllegalServantException             | ArgumentNullException (C#), std::invalid_argument (C++) | |
```

File: cpp/include/Ice/LocalExceptions.h [hunk 11]
```diff
@@ -323,897 +280,636 @@ namespace Ice
...
-        {
-        }
+        SyscallException(const char* file, int line, std::string messagePrefix, ErrorCode error);
 
         /**
-         * Obtains a tuple containing all of the exception's data members.
-         * @return The data members in a tuple.
+         * Gets the error number describing the system exception. For C++ and Unix, this is equivalent to
+         * <code>errno</code>. For C++ and Windows, this is the value returned by <code>GetLastError()</code> or
+         * <code>WSAGetLastError()</code>.
```
Comment (externl) [inline cpp/include/Ice/LocalExceptions.h:387]:
```suggestion
         * Gets the error number describing the system exception. On Unix, this is equivalent to
         * <code>errno</code>. On Windows, this is the value returned by <code>GetLastError()</code> or
         * <code>WSAGetLastError()</code>.
```

File: matlab/src/Util.cpp [hunk 12]
```diff
@@ -373,29 +421,17 @@ IceMatlab::convertException(const std::exception_ptr exc)
             createIdentity(e.id()),
             createStringFromUTF8(e.facet()),
             createStringFromUTF8(e.operation()),
+
             createStringFromUTF8(errID),
             createStringFromUTF8(e.what())};
```
Comment (externl) [inline matlab/src/Util.cpp:426]:
```suggestion
            createIdentity(e.id()),
            createStringFromUTF8(e.facet()),
            createStringFromUTF8(e.operation()),
            createStringFromUTF8(errID),
            createStringFromUTF8(e.what())};
```

**Review by externl — APPROVED** [review]
Looks good to me!

**Review by pepone — APPROVED** [review]

File: csharp/src/Ice/ConnectionI.cs [hunk 13]
```diff
@@ -1543,7 +1543,7 @@ internal void idleCheck(TimeSpan idleTimeout)
                         $"connection aborted by the idle check because it did not receive any bytes for {idleTimeoutInSeconds}s\n{_transceiver.toDetailedString()}");
                 }
 
-                setState(StateClosed, new ConnectionIdleException($"Connection aborted by the idle check because it did not receive any bytes for {idleTimeoutInSeconds}s."));
+                setState(StateClosed, new ConnectionIdleException($"The connection was aborted by the idle check because it did not receive any bytes for {idleTimeoutInSeconds}s."));
```
[the remark this answers is not in the document]
Comment (bernardnormier) [reply csharp/src/Ice/ConnectionI.cs:1546]: Yes, will do in a follow-up C# PR.

File: cpp/include/Ice/IconvStringConverter.h [hunk 14]
```diff
@@ -37,28 +37,13 @@ namespace Ice
     class ICE_API IconvInitializationException final : public LocalException
```
[the remark this answers is not in the document]
Comment (bernardnormier) [reply cpp/include/Ice/IconvStringConverter.h]: Fixed. FeatureNotSupportedException was actually a better match.

**Pull request merged**